### DISCUSSION TITLE
Security Issue

### DISCUSSION THREAD

**Discussion by FlyingNimbusCloud** [christiaangoossens/hass-oidc-auth#11] [Ideas]
I know this is alpha and there's lots to do. But one thing I wanted to point out is that the One Time Password is only a 6 digit pin. An attacker could potentially guess a 6 digit number and gain access during the 5 minute window after the OTP was generated.

**Comment by christiaangoossens** [comment]
I thought about it, wanting to make a balance between easy to type into your mobile device and security, and I think 6 digits is a pretty acceptable compromise.

However, I will add some additional rate limiting/slowing down the requests in the next release to make sure. 
Hope this clarifies the choice for 6 digit pins! Thanks for your message.

---

Here's my reasoning. In most cases, the code will only be live for a minute or less, especially if you click the direct sign in button, as it expires after the first use. So, let's assume we are talking about the worst (normal) case here:

- You have logged in, thus generating a code, but you will never use the code, thus they have the full 5 minutes to guess it
- You are running HA on a pretty fast computer, not on an old Raspberry Pi or so
- You know that the user is using this integration and thus that it must be 6 digits, not alphanumeric etc (and we don't offer an option to configure it)
- You have access to the HA instance through the internet, fully without any type of rate limiting from proxies you have configured
- They have not enabled the `login_attempts_threshold` (https://www.home-assistant.io/integrations/http/) which would block you after a certain amount of attempts. This is not set by default, thus having banning disabled.
- The attacker has some scouting, knows the flow ID & has written code to get the HA token upon first successful try

You will need to try 10<sup>6</sup> combinations (0-9 for 6 digits) in total to try all possible options. But likely you will need much less (like a sixth) with a bit of luck (and a random uniform code distribution).

Every wrong attempt currently (v0.4.0-alpha) registers a wrong login attempt: `homeassistant  | { DATE TIME } WARNING (MainThread) [homeassistant.components.http.ban] Login attempt or request with invalid authentication from {IP}. Requested URL: '/auth/login_flow/{FLOW ID}'. ({ USER AGENT })`. If you **had** enabled login_attempts_threshold, the malicious IP would be banned soon.

Home Assistant by default does not enable any rate limiting or banning. It only relies on the password attempt taking long enough (they force it to always take 200ms from my testing). _Sidenote: I would argue the HA default login is actually much less secure as you have an infinite amount of time to try as the credentials don't change or expire and they don't enable banning by default._

For my computer, an attempt for the OIDC code takes about 15ms. Let's continue with 10ms per request to be a bit more generous (on a Raspberry Pi, this will likely be closer to 30ms). That means that in 300 seconds (300000 ms) you can try 30k codes. That's 3% of all possible codes, which is still a lot less than the ~166k codes that I think you would reasonably need to try to get in (1/6th of all codes).

If we also throttle back to 200ms like the native HA integration, you would only be able to try at most 1500 codes. That's only 0.15% of all possible codes, assuming you start immediately when the code is generated.

**Reply by FlyingNimbusCloud** [reply]
Thanks this is really well thought out. I am new to homeassistant and didn't realize it has built in rate limiting and ip banning. I was actually looking into fail2ban config for reading logs on homeassistant and implementing my own temp bans on ips. I think it might be good to at least reference the homeassistant docs that describe these features in this repository's documentation as a recommended security measure to protect against OTP brute force attacks. 

I am going to fix my container and implement these things right away.

Thanks again! So stoked about this integration.

**Reply by christiaangoossens** [reply]
@FlyingNimbusCloud Definitely enable the login_attempts_threshold and configure your proxies correctly such that the right IPs get banned automatically. The default settings are actually pretty bad at protecting its username/password as you get infinite attempts, even without this integration.

I will add some links to the README later, thanks for the feedback.

**Comment by christiaangoossens** [comment]
Fixed with alpha version 0.4.1 (https://github.com/christiaangoossens/hass-oidc-auth/releases/tag/v0.4.1-alpha), which slows it down to the same request duration as the normal HA provider.
